Add blog post for Parquet and Feather 3/3 (#2801)
tenzir/vast: web/blog/parquet-and-feather-data-engineering-woes/index.md

---

> draft: true

Reviewer: This should be removed.

> title: "Parquet & Feather: Data Engineering Woes"

Reviewer: Consistency nit: We don't quote any other YAML strings in the frontmatter, so why this one?

Reviewer 2: The `&`?

Reviewer: Ah—I thought that only started an anchor when followed by something that's not a space.

authors:
  - dispanser
  - mavam
date: 2023-01-10

> image: arrow-schema-conversion.light.png

Reviewer: The carousel on the frontpage of the docs checks if this path begins with a slash, and otherwise refuses to render the image. Not sure if the bug is in the carousel or here.

Reviewer 2: I removed this. Unless an image is in `/static/img` and can be referenced via `/img/...`, the image will not be properly displayed because the name gets mangled to something internal.

own [type system](https://vast.io/docs/understand-vast/data-model/type-system)
relying on [Arrow extension
types](https://arrow.apache.org/docs/format/Columnar.html#extension-types) to
express domain-specific concepts like IP addresses, subnets, or enumerations. We
hit a few places where the Arrow C++ implementation does not support all VAST
types directly. It's trickier than we thought, as we see next.

## Row Groups

In Apache Parquet, a [row group](https://parquet.apache.org/docs/concepts/) is a
subset of a Parquet file that's itself written in a columnar fashion. Smaller
row groups allow for higher granularity in reading parts of an individual file,
at the expense of a potentially increased file size due to less optimal
encoding. In VAST, we send around [batches of
data](https://vast.io/docs/setup-vast/tune) that are considerably smaller than
what a recommended Parquet file size would look like. A typical Parquet file
size recommendation is 1GB, which translates to 5–10GB of data in memory when
reading the entire file. To produce files sized in this order of magnitude, we
planned to use individual row groups, each of which aligned with the size of our

> Arrow record batches that comprise 2^16 events occopying a few MBs.

Reviewer: ```suggestion
Arrow record batches that comprise 2<sup>16</sup> events occupying a few MBs.
```

into the Parquet file metadata, this doesn't seem to play well in concert with
extension types. As a result, a record batch written to and then read from a
Parquet file no longer adheres to the same schema! 

This bit us in the following scenarios.

### VAST Enumerations

VAST comes with an enumeration type that represents a fixed mapping of strings
to numeric values, where the mapping is part of the type metadata. We represent
enums as extension types wrapping an Arrow dictionary of strings backed by
unsigned 8-bit integers. On read, Arrow turns these 8-bit index values into
32-bit values, which is not compatible with our extension type definition, so
the extension type wrapping is lost. The diagram below illustrates this issue.

![Arrow Schema Conversion #width500](arrow-schema-conversion.light.png#gh-light-mode-only)
![Arrow Schema Conversion #width500](arrow-schema-conversion.dark.png#gh-dark-mode-only)

### Extension Types inside Maps

Both our address type and subnet type extensions are lost if they occur in
nested records. For example, a map from a VAST address to a VAST enumeration of
the following Arrow type is not preserved:

```

> map<extension<vast.address>,  extension<vast.enumeration>>

Reviewer: ```suggestion
map<extension<vast.address>, extension<vast.enumeration>>
```

stack overflow post](https://stackoverflow.com/q/64383029) has the solution: if
we define an explicit schema, Spark happily converts our column into a signed
type.

```scala
val schema = StructType(
  Array(
    StructField("event",
      StructType(
        Array(
          StructField("c", LongType))))))
```

Finally, it works!

```
scala> spark.read.schema(schema).parquet(<file>).show()
+-----+
|event|
+-----+
| {13}|
+-----+
```

We were able to read VAST data in Spark, but it's not an easy and out-of-the-box
experience we were hoping for. It turns out that different tools don't always
support all the data types, and additional effort is required to integrate with
the big players in the Parquet ecosystem.

## Conclusion

We love Apache Arrow—it's a cornerstone of our system, and we'd be in much
worse shape without it. We use it everywhere from the storage layer (using
Feather and Parquet) to the data plane (where we are passing around Arrow record
batches).

However, as VAST uses a few less common Arrow features we sometimes stumble over
some of the rougher edges. We're looking forward to fixing some of these things

> upstream, but sometimes you just need a quick solution to help our users..

Reviewer: ```suggestion
upstream, but sometimes you just need a quick solution to help our users.
```
